include/drivers/reset.h

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022 Andrei-Edward Popa <andrei.popa105@yahoo.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public Reset Controller driver APIs
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+#define ZEPHYR_INCLUDE_DRIVERS_RESET_H_
+
+/**
+ * @brief Reset Controller Interface
+ * @defgroup reset_controller_interface Reset Controller Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** Reset controller device configuration. */
+struct reset_dev_config {
+	/** Reset controller device. */
+	const struct device *dev;
+	/** Reset line. */
+	uint32_t id;
+};
+
+/**
+ * @brief Helper macro to initialize reset controller device config.
+ *
+ * @param inst Instance number.
+ * @param id Reset line.
+ */
+#define RESET_DEV_CONFIG_INIT(inst, id)						\
+	{									\
+		.dev = DEVICE_DT_GET(DT_INST_RESET_CTLR(inst)),			\
+		.id = DT_INST_RESET_CELL(inst, id),				\
+	}
```

Comment:
I'm a bit puzzled about this. Who is going to use the `struct reset_dev_config` and how? Take for example https://github.com/zephyrproject-rtos/zephyr/blob/8a8bc392dc93a78ec3a84ba68811b46a3cc73ffa/include/drivers/mbox.h#L92-L98
In this case we have added some helper functions (see the `mbox_*` functions) that leverage that struct. But in your case how you are supposed to use this struct and why is it there if there is no user?

Reply:
I tried to create this struct for getting the reset controller device and its line into some device driver (you can see an example of this in drivers/serial/uart_rpi_pico.c) and if you want to reset the device you can say something like this:
reset_reset(reset->dev, reset->line);
In the line below, 'reset' is a pointer to a reset_dev_config.
I'm not sure how to do it in another way, but if you have any idea I will gladly receive it.

Reply:
> if you want to reset the device you can say something like this:
> reset_reset(reset->dev, reset->line);

yes, but that makes sense if you have a wrapper at API level that can unpack that for you, for example `reset((struct reset_dev_config *) conf)`. In your case that wrapper is not being defined at API level so this struct is simply unused (I mean, the same struct can be created by the driver if it is going to be used right?)

Reply:
Okay, yes, I think I understand it, I will try to create some wrappers.

Reply:
@carlocaione I have already a syscall called reset_reset as you saw, and this takes a dev pointer and an id and now I want to create another syscall that takes a reset_dev_config. How can I call it (I'm talking about its name)?

Reply:
no, do not create more syscalls. All your syscalls are already taking a dev pointer and an id, why don't you simply change the syscalls you have already to take a `struct reset_dev_config` instead?

Reply:
Okay, that's cool.

Reply:
you should be using `device *` and not this struct. The reason is that `struct device` is a kernel object and because of this you can grant or revoke access per thread. You can also go additional validations like if the type of the device is right ...

Reply:
@ceolin What about the MBOX device? It is using a structure like I did. Maybe @carlocaione or @gmarull can clarify this? I'm not sure how to do it in another way and some suggestions would be useful.

Reply:
@ceolin you can still do that, see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mbox/mbox_handlers.c